error: storage size of ‘par’ isn’t known

[build] /home/soham/Documents/music_payer_c/lib/libsndfile/programs/sndfile-play.c: In function ‘sndio_play’:
[build] /home/soham/Documents/music_payer_c/lib/libsndfile/programs/sndfile-play.c:677:25: error: storage size of ‘par’ isn’t known
[build]   677 |         struct sio_par  par ;
[build]       |                         ^~~
[build] [ 56%] Built target sndfile-loopify
[build] /home/soham/Documents/music_payer_c/lib/libsndfile/programs/sndfile-play.c:695:28: warning: implicit declaration of function ‘sio_open’; did you mean ‘sf_open’? [-Wimplicit-function-declaration]
[build]   695 |                 if ((hdl = sio_open (NULL, SIO_PLAY, 0)) == NULL)
[build]       |                            ^~~~~~~~
[build]       |                            sf_open
[build] /home/soham/Documents/music_payer_c/lib/libsndfile/programs/sndfile-play.c:695:44: error: ‘SIO_PLAY’ undeclared (first use in this function)
[build]   695 |                 if ((hdl = sio_open (NULL, SIO_PLAY, 0)) == NULL)
[build]       |                                            ^~~~~~~~
[build] /home/soham/Documents/music_payer_c/lib/libsndfile/programs/sndfile-play.c:695:44: note: each undeclared identifier is reported only once for each function it appears in

--- a/configure.ac
+++ b/configure.ac
@@ -477,23 +477,27 @@ dnl ============================================================================
 dnl  Check for Sndio.
 
 HAVE_SNDIO_H=0
-AS_IF([test "x$enable_sndio" != "xno"], [
-		PKG_CHECK_MODULES([SNDIO], [sndio], [
-				HAVE_SNDIO_H=1
-				ac_cv_sndio="yes"
-			], [
-				ac_cv_sndio="no"
-			])
-
-		AS_IF([test "x$ac_cv_sndio" = "xno"], [
-				AS_IF([test "x$enable_sndio" = "xyes"], [
-						dnl explicitly passed --enable-sndio, hence error out loud and clearly
-						AC_MSG_ERROR([You explicitly requested sndio support, but sndio could not be found!])
-                                        ], [
-						dnl did not explicitly pass --enable-sndio, relying on default automagic on
-						enable_sndio="no (auto)"
+AS_IF([test "x$alsa_works" != "xyes"], [
+		AS_IF([test "x$enable_sndio" != "xno"], [
+				PKG_CHECK_MODULES([SNDIO], [sndio], [
+						HAVE_SNDIO_H=1
+						ac_cv_sndio="yes"
+					], [
+						ac_cv_sndio="no"
+					])
+		
+				AS_IF([test "x$ac_cv_sndio" = "xno"], [
+						AS_IF([test "x$enable_sndio" = "xyes"], [
+								dnl explicitly passed --enable-sndio, hence error out loud and clearly
+								AC_MSG_ERROR([You explicitly requested sndio support, but sndio could not be found!])
+							], [
+								dnl did not explicitly pass --enable-sndio, relying on default automagic on
+								enable_sndio="no (auto)"
+							])
 					])
 			])
+	], [
+		AC_MSG_WARN([You have requested ALSA already, disable ALSA if you want to use Sndio.])
 	])
 
 AC_DEFINE_UNQUOTED([HAVE_SNDIO_H], [${HAVE_SNDIO_H}], [Set to 1 if <sndio.h> is available.])

--- a/programs/sndfile-play.c
+++ b/programs/sndfile-play.c
@@ -52,26 +52,25 @@
 	#define ALSA_PCM_NEW_SW_PARAMS_API
 	#include <alsa/asoundlib.h>
 	#include <sys/time.h>
-#endif
-
-#if defined (__ANDROID__)
-
-#elif defined (__linux__) || defined (__FreeBSD_kernel__) || defined (__FreeBSD__) || defined (__riscos__)
-	#include 	<fcntl.h>
-	#include 	<sys/ioctl.h>
-	#include 	<sys/soundcard.h>
 
 #elif HAVE_SNDIO_H
 	#include <sndio.h>
+
+#elif defined (__ANDROID__)
+
+#elif defined (__linux__) || defined (__FreeBSD_kernel__) || defined (__FreeBSD__) || defined (__riscos__)
+	#include <fcntl.h>
+	#include <sys/ioctl.h>
+	#include <sys/soundcard.h>
+
+#elif (OS_IS_WIN32 == 1)
+	#include <windows.h>
+	#include <mmsystem.h>
 
 #elif (defined (sun) && defined (unix)) || defined(__NetBSD__)
 	#include <fcntl.h>
 	#include <sys/ioctl.h>
 	#include <sys/audioio.h>
-
-#elif (OS_IS_WIN32 == 1)
-	#include <windows.h>
-	#include <mmsystem.h>
 
 #endif
 
@@ -349,13 +348,71 @@ alsa_write_float (snd_pcm_t *alsa_dev, float *data, int frames, int channels)
 	return total ;
 } /* alsa_write_float */
 
-#endif /* HAVE_ALSA_ASOUNDLIB_H */
+/*------------------------------------------------------------------------------
+**	Sndio.
+*/
+
+#elif HAVE_SNDIO_H
+
+static void
+sndio_play (int argc, char *argv [])
+{	struct sio_hdl	*hdl ;
+	struct sio_par	par ;
+	short buffer [BUFFER_LEN] ;
+	SNDFILE	*sndfile ;
+	SF_INFO	sfinfo ;
+	int		k, readcount ;
+
+	for (k = 1 ; k < argc ; k++)
+	{	printf ("Playing %s\n", argv [k]) ;
+		if (! (sndfile = sf_open (argv [k], SFM_READ, &sfinfo)))
+		{	puts (sf_strerror (NULL)) ;
+			continue ;
+			} ;
+
+		if (sfinfo.channels < 1 || sfinfo.channels > 2)
+		{	printf ("Error : channels = %d.\n", sfinfo.channels) ;
+			continue ;
+			} ;
+
+		if ((hdl = sio_open (NULL, SIO_PLAY, 0)) == NULL)
+		{	fprintf (stderr, "open sndio device failed") ;
+			return ;
+			} ;
+
+		sio_initpar (&par) ;
+		par.rate = sfinfo.samplerate ;
+		par.pchan = sfinfo.channels ;
+		par.bits = 16 ;
+		par.sig = 1 ;
+		par.le = SIO_LE_NATIVE ;
+
+		if (! sio_setpar (hdl, &par) || ! sio_getpar (hdl, &par))
+		{	fprintf (stderr, "set sndio params failed") ;
+			return ;
+			} ;
+
+		if (! sio_start (hdl))
+		{	fprintf (stderr, "sndio start failed") ;
+			return ;
+			} ;
+
+		while ((readcount = sf_read_short (sndfile, buffer, BUFFER_LEN)))
+			sio_write (hdl, buffer, readcount * sizeof (short)) ;
+
+		sio_close (hdl) ;
+		} ;
+
+	return ;
+} /* sndio_play */
+
+#elif defined (__ANDROID__)
 
 /*------------------------------------------------------------------------------
 **	Linux/OSS functions for playing a sound.
 */
 
-#if !defined (__ANDROID__) && (defined (__linux__) || defined (__FreeBSD_kernel__) || defined (__FreeBSD__) || defined (__riscos__))
+#elif defined (__linux__) || defined (__FreeBSD_kernel__) || defined (__FreeBSD__) || defined (__riscos__)
 
 static	int	opensoundsys_open_device (int channels, int srate) ;
 
@@ -466,8 +523,6 @@ opensoundsys_open_device (int channels, int srate)
 	return 	fd ;
 } /* opensoundsys_open_device */
 
-#endif /* __linux__ */
-
 /*------------------------------------------------------------------------------
 **	Mac OS X functions for playing a sound.
 */
@@ -482,7 +537,7 @@ opensoundsys_open_device (int channels, int srate)
 **  point to data instead of short*. It plain sucks!
 */
 
-#if (OS_IS_WIN32 == 1)
+#elif (OS_IS_WIN32 == 1)
 
 #define	WIN32_BUFFER_LEN	(1 << 15)
 
@@ -663,73 +718,11 @@ win32_play (int argc, char *argv [])
 
 } /* win32_play */
 
-#endif /* Win32 */
-
-/*------------------------------------------------------------------------------
-**	Sndio.
-*/
-
-#if HAVE_SNDIO_H
-
-static void
-sndio_play (int argc, char *argv [])
-{	struct sio_hdl	*hdl ;
-	struct sio_par	par ;
-	short buffer [BUFFER_LEN] ;
-	SNDFILE	*sndfile ;
-	SF_INFO	sfinfo ;
-	int		k, readcount ;
-
-	for (k = 1 ; k < argc ; k++)
-	{	printf ("Playing %s\n", argv [k]) ;
-		if (! (sndfile = sf_open (argv [k], SFM_READ, &sfinfo)))
-		{	puts (sf_strerror (NULL)) ;
-			continue ;
-			} ;
-
-		if (sfinfo.channels < 1 || sfinfo.channels > 2)
-		{	printf ("Error : channels = %d.\n", sfinfo.channels) ;
-			continue ;
-			} ;
-
-		if ((hdl = sio_open (NULL, SIO_PLAY, 0)) == NULL)
-		{	fprintf (stderr, "open sndio device failed") ;
-			return ;
-			} ;
-
-		sio_initpar (&par) ;
-		par.rate = sfinfo.samplerate ;
-		par.pchan = sfinfo.channels ;
-		par.bits = 16 ;
-		par.sig = 1 ;
-		par.le = SIO_LE_NATIVE ;
-
-		if (! sio_setpar (hdl, &par) || ! sio_getpar (hdl, &par))
-		{	fprintf (stderr, "set sndio params failed") ;
-			return ;
-			} ;
-
-		if (! sio_start (hdl))
-		{	fprintf (stderr, "sndio start failed") ;
-			return ;
-			} ;
-
-		while ((readcount = sf_read_short (sndfile, buffer, BUFFER_LEN)))
-			sio_write (hdl, buffer, readcount * sizeof (short)) ;
-
-		sio_close (hdl) ;
-		} ;
-
-	return ;
-} /* sndio_play */
-
-#endif /* sndio */
-
 /*------------------------------------------------------------------------------
 **	Solaris.
 */
 
-#if (defined (sun) && defined (unix)) || defined(__NetBSD__)
+#elif (defined (sun) && defined (unix)) || defined(__NetBSD__)
 
 static void
 solaris_play (int argc, char *argv [])
@@ -830,31 +823,35 @@ main (int argc, char *argv [])
 		return 1 ;
 		} ;
 
-#if defined (__ANDROID__)
-	puts ("*** Playing sound not yet supported on Android.") ;
-	puts ("*** Please feel free to submit a patch.") ;
-	return 1 ;
-#elif defined (__linux__)
-	#if HAVE_ALSA_ASOUNDLIB_H
-		if (access ("/proc/asound/cards", R_OK) == 0)
-			alsa_play (argc, argv) ;
-		else
-	#endif
-		opensoundsys_play (argc, argv) ;
-#elif defined (__FreeBSD_kernel__) || defined (__FreeBSD__) || defined (__riscos__)
-	opensoundsys_play (argc, argv) ;
+#if HAVE_ALSA_ASOUNDLIB_H
+	if (access ("/proc/asound/cards", R_OK) == 0)
+		alsa_play (argc, argv) ;
+	else
+		return EXIT_FAILURE ;
+
 #elif HAVE_SNDIO_H
 	sndio_play (argc, argv) ;
-#elif (defined (sun) && defined (unix)) || defined(__NetBSD__)
-	solaris_play (argc, argv) ;
+
+#elif defined (__ANDROID__)
+	puts ("*** Playing sound not yet supported on Android.") ;
+	puts ("*** Please feel free to submit a patch.") ;
+	return EXIT_FAILURE ;
+
+#elif defined (__linux__) || defined (__FreeBSD_kernel__) || defined (__FreeBSD__) || defined (__riscos__)
+	opensoundsys_play (argc, argv) ;
+
 #elif (OS_IS_WIN32 == 1)
 	win32_play (argc, argv) ;
+
+#elif (defined (sun) && defined (unix)) || defined(__NetBSD__)
+	solaris_play (argc, argv) ;
+
 #else
 	puts ("*** Playing sound not supported on this platform.") ;
 	puts ("*** Please feel free to submit a patch.") ;
-	return 1 ;
+	return EXIT_FAILURE ;
 #endif
 
-	return 0 ;
+	return EXIT_SUCCESS ;
 } /* main */
 

--- a/src/sfconfig.h
+++ b/src/sfconfig.h
@@ -41,6 +41,10 @@
 #define HAVE_ALSA_ASOUNDLIB_H 0
 #endif
 
+#ifndef HAVE_SNDIO_H
+#define HAVE_SNDIO_H 0
+#endif
+
 #ifndef HAVE_BYTESWAP_H
 #define HAVE_BYTESWAP_H 0
 #endif

